add cloudforms inventory script

mrjoshuap opened the pull request:
[cloudforms--list.json.txt](https://github.com/ansible/ansible/files/412082/cloudforms--list.json.txt)
[cloudforms--host.json.txt](https://github.com/ansible/ansible/files/412086/cloudforms--host.json.txt)
##### ISSUE TYPE
- Feature Pull Request
##### COMPONENT NAME

<!--- Name of the plugin/module/task -->

Contributed inventory for Red Hat CloudForms / ManageIQ
##### ANSIBLE VERSION

<!--- Paste verbatim output from “ansible --version” between quotes below -->

```
ansible 2.1.0.0
  config file = /etc/ansible/ansible.cfg
  configured module search path = Default w/o overrides
```
##### SUMMARY

<!--- Describe the change, including rationale and design decisions -->

This is a new contributed inventory script for Red Hat CloudForms / ManageIQ.  I tested with Red Hat CloudForms 4.1.  See attached output examples.

<!-- Paste verbatim command output below, e.g. before and after your change -->

```
usage: cloudforms.py [-h] [--list] [--host HOST] [--refresh-cache] [--debug]

Produce an Ansible Inventory file based on CloudForms managed VMs

optional arguments:
  -h, --help       show this help message and exit
  --list           List instances (default: True)
  --host HOST      Get all the variables about a specific instance
  --refresh-cache  Force refresh of cache by making API requests to cloudforms
                   (default: False - use cache files)
  --debug          Show debug output while running (default: False)
```

based on the foreman inventory script, features:
- cached results (default 600 seconds)
- paginated host results (default 100 hosts)
- ssl verification (default True)
- arguments to flush cache and run in debug mode

mrjoshuap committed:
add cloudforms inventory script

based on the foreman inventory script, features:

* cached results (default 600 seconds)
* paginated host results (default 100 hosts)
* ssl verification (default True)
* arguments to flush cache and run in debug mode

nitzmahone commented on contrib/inventory/cloudforms.py:
```diff
@@ -0,0 +1,363 @@
+            for hostname in self.cache:
+                self.inventory['_meta']['hostvars'][hostname] = {
+                    'cloudforms': self.cache[hostname],
+                    'cloudforms_params': self.params[hostname],
```
Why the two dicts? Seems like some stuff is duplicated between cloudforms and cloudforms_params- would it be more user-friendly to squash to a single "cloudforms_facts" dict or something?

nitzmahone commented:
Sweet- just the one question about the two sub-dicts, otherwise LGTM!

mrjoshuap commented:
There is no technical reason they can't be squashed into a single dict; this was based on the foreman.py inventory script that shipped with Ansible Tower 3.0.1, which used the second for host group parameters.  Let me rework it with the suggestion and update this PR.

mrjoshuap committed:
suggested rework

* removed second cache / dict with duplicate info
* added purge_actions configuration option to remove the actions from a host (defaults to False)
* added prefer_ip_address configuration option so give the option of using ip address instead of name (defaults to True)
* removed self variables — just use the arguments directly
* added --pretty command line option to pretty print results
* renamed _resolve_params to _resolve_host

mrjoshuap commented:
Reworked to squash into a single dict.  Additionally:
- removed second cache / dict with duplicate info
- added purge_actions configuration option to remove the actions from a host (defaults to False)
- added prefer_ip_address configuration option so give the option of using ip address instead of name (defaults to True)
- removed self variables — just use the arguments directly
- added --pretty command line option to pretty print results
- renamed _resolve_params to _resolve_host

New output attached.

[cloudforms--host.json.txt](https://github.com/ansible/ansible/files/415494/cloudforms--host.json.txt)
[cloudforms--list.json.txt](https://github.com/ansible/ansible/files/415495/cloudforms--list.json.txt)

mrjoshuap commented:
Also, --help updated:

```
usage: cloudforms.py [-h] [--list] [--host HOST] [--pretty] [--refresh-cache]
                     [--debug]

Produce an Ansible Inventory file based on CloudForms managed VMs

optional arguments:
  -h, --help       show this help message and exit
  --list           List instances (default: True)
  --host HOST      Get all the variables about a specific instance
  --pretty         Pretty print JSON output (default: False)
  --refresh-cache  Force refresh of cache by making API requests to CloudForms
                   (default: False - use cache files)
  --debug          Show debug output while running (default: False)
```

jlaska commented on contrib/inventory/cloudforms.py:
```diff
@@ -0,0 +1,351 @@
+                        print "Found tag [%s] for host which will be mapped to [%s]" % (group['name'], safe_key)
+
+            full_host = self._resolve_host(host)
+            if host.has_key('ipaddresses'):
```
Can you use `if 'ipaddresses' in host` instead?  We should use this where ever you have `has_key()`

mrjoshuap replied:
I'll rework both of them as suggested and catch any others.

jlaska commented on contrib/inventory/cloudforms.py:
```diff
@@ -0,0 +1,351 @@
+                print "Host named [%s] will use [%s] as dns_name" % (host['name'], dns_name)
+
+            # Create ansible groups for hostgroup, location and organization
+            if host.has_key('tags'):
```
Suggest changing to `if 'tags' in host:`

mrjoshuap replied:
done 👍

jlaska commented on contrib/inventory/cloudforms.py:
```diff
@@ -0,0 +1,351 @@
+                    print "Skipping %s because power_state = %s" % (host['name'], host['power_state'])
+                continue
+
+            if self.cloudforms_prefer_ip_address and host.has_key('ipaddresses') and host['ipaddresses'][0]:
```
Instead of changing the `name`, can we set `ansible_ssh_host` to the discovered `ipaddress`?

jlaska replied:
Also, this results in a traceback for me.  I think you want to do ...

``` python
if 'ipaddresses' in host and isinstance(host['ipaddresses'], list) and host['ipaddresses']:
```

mrjoshuap replied:
this is actually not needed since it will keep the name and specify `ansible_ssh_host` as the first IP.

mrjoshuap commented:
I'm addressing all of those concerns and adapting it to catch more exceptions.

mrjoshuap committed:
implement suggestions

* removed not used import
* added warnings to help debug connection issues
* renamed self.cache to self.hosts for clarity
* now will use the first ip address as ansible_ssh_host
* flipped default for prefer_ip_address config option to false - preserve name, and specify ansible_ssh_host as ip address
* added checks and warnings to configuration options, sane defaults for all except required:
** `url` - the first part of the cloudforms server url (https://cfme.example.com)
** `username`  - the cloudforms username to log in with
** `password` - the password for the cloudforms user specified
* removed redundant call to fetch host information (since we’re paging results, no need to split the calls)
* added warning for unexpected responses from CloudForms
* debug for returned sting now prints the string instead of forcing to JSON
* removed no longer needed methods to fetch host information
* using ‘key in list’ instead of ‘list.has_key(key)’
* correctly formatted groups and allowed nested groups
* now create groups for `location`, `type` and `vendor`, with appropriate sub-groups and children
* made to_safe honor config option to clean group names for ansible consumption

mrjoshuap committed:
remove prefer_ip_address configuration option

no longer needed since we will specify `ansible_ssh_host` as the returned ip address.

mrjoshuap committed:
removed dns_name

no longer needed, will preserve `host[name]` as name in Ansible.

mrjoshuap commented:
All issues and suggestions addressed, needs reviewed and tested by others :-)

```
usage: cloudforms.py [-h] [--list] [--host HOST] [--pretty] [--refresh-cache]
                     [--debug]

Produce an Ansible Inventory file based on CloudForms managed VMs

optional arguments:
  -h, --help       show this help message and exit
  --list           List instances (default: True)
  --host HOST      Get all the variables about a specific instance
  --pretty         Pretty print JSON output (default: False)
  --refresh-cache  Force refresh of cache by making API requests to CloudForms
                   (default: False - use cache files)
  --debug          Show debug output while running (default: False)
```

Also updated outputs:

[cloudforms_py--host.json.txt](https://github.com/ansible/ansible/files/423716/cloudforms_py--host.json.txt)
[cloudforms_py--list.json.txt](https://github.com/ansible/ansible/files/423717/cloudforms_py--list.json.txt)

mrjoshuap commented:
Sidebar:

Can anyone confirm if group names need to be sanitized?  I have checked and searched but not found anything definitive.  For example, one of the tags returned is named **/managed/folder_path_yellow/datacenters** and a type is returned as **ManageIQ::Providers::Redhat::InfraManager::Vm** ... I believe slashes and colons will prevent Ansible from using them as groups or worse?

I added a configuration option to make it optional and defaulted it to True because I wasn't sure about the character validity in Ansible...

They are now added (as groups) as:
- _managed_folder_path_yellow_datacenters
- ManageIQ__Providers__Redhat__InfraManager__Vm

I could argue that at least for tags that I might want to make group -> subgroup split on the slashes... others are not as straight forward.

mrjoshuap commented:
It is not a good idea to split tags on the slashes (/) ... the last sub-group key can be duplicated in multiple parents squashing all hosts into a single group and wreaking havoc upon unsuspecting innocents.

mrjoshuap commented:
I don't have enough data to confirm my previous comment re: splitting tags on /... I created a new branch where this could be tested but not act as a blocker for this PR.

https://github.com/mrjoshuap/ansible/tree/cloudforms_inventory_nested_tags

mrjoshuap committed:
Merge branch 'devel' of github.com:ansible/ansible into cloudforms_inventory

jlaska commented:
@mrjoshuap in your sample JSON output, you incude the `actions` key for each host.  Not a show stopper, but I wonder if that data is useful to ansible?  Should it be excluded in the hostvars?

jlaska commented on contrib/inventory/cloudforms.py:
```diff
@@ -0,0 +1,400 @@
+        try:
+            results = json.loads(ret.text)
+        except ValueError:
+            warnings.warn("Unexpected response from {0} ({1}): {2}".format(self.cloudforms_url, r.status_code, r.reason))
```
heads up, `r` is undefined.

mrjoshuap replied:
Whoops!  silly me.  On that now.

jlaska replied:
Oh, I see you've made the `actions` configurable.  Disregard that comment.

jlaska commented:
Nit pick, but running this through `flake8` will highlight a few other minor gotchas.

mrjoshuap commented:
I agree with regards to actions; I made it configurable and default it to True; should I default it to False?  Also, I will utilize the flake8 and see what it says.

mrjoshuap committed:
purge actions from hostvars

changed purge_actions to True

mrjoshuap committed:
flake8 suggestion for whitespace

mrjoshuap committed:
fix undefined r variable in warning output

use the correct ret variable

mrjoshuap commented:
flake8 returns all clear with exception of long lines, some of which I'm not sure the best way to resolve.  Are we extremely picky on that front?

mrjoshuap committed:
Default purge_actions to True

We probably don’t need them, but it is configurable, so just default to remove them.

mrjoshuap committed:
Add configuration option to nest cloudforms tags

disabled by default, the nest_tags option will expand cloudforms tags into a nested group/subgroup structure.  Otherwise, it will use the whole tag name.

mrjoshuap committed:
added purging the actions

removed in previous clean up in error.

mrjoshuap commented:
For some reason, I removed the purge_actions method and forgot to add it back in.  I fixed that.  Also, added a nest_tags option so that CloudForms tags would show up as a nested list.  I split them based on the / in the tag name.  It is disabled by default because my sample size is vary small and there might be problems when a node with the same name appears in multiple tags.

mrjoshuap commented:
Here's some sample outputs....

[cloudforms_py--host.json.txt](https://github.com/ansible/ansible/files/425403/cloudforms_py--host.json.txt)
[cloudforms_py--list--nested-tags.json.txt](https://github.com/ansible/ansible/files/425404/cloudforms_py--list--nested-tags.json.txt)
[cloudforms_py--list.json.txt](https://github.com/ansible/ansible/files/425405/cloudforms_py--list.json.txt)

jlaska commented:
> flake8 returns all clear with exception of long lines, some of which I'm not sure the best way to resolve. Are we extremely picky on that front?

I'm not, I don't believe @nitzmahone is either.

jlaska commented on contrib/inventory/cloudforms.py:
```diff
@@ -0,0 +1,460 @@
+                        safe_key = self.to_safe(group['name'])
+                        if safe_key:
+                            if self.args.debug:
+                                print "Adding sub-group '%s' to parent 'tags'" % safe_tag_name
```
Heads  up, I think `safe_tag_name` is undefined here

mrjoshuap replied:
you are correct sir, my bad, I missed the output from flake8!  /sigh

mrjoshuap committed:
fixed undefined variable

specified the correct variable for logging.

nitzmahone commented:
Nope- we're trying to tighten up pep/flake kinds of things in the core codebase going forward, but this being contrib, we probably don't need to be too militant. :)

jlaska commented:
@nitzmahone looks good to me, thanks @mrjoshuap for all the work!
